clippy_lints/src/map_unit_fn.rs

```diff
@@ -0,0 +1,263 @@
+                    // for all cases (multi-line closures specifically)
+                    //
+                    // We do not attempt to build a suggestion for those right now.
+                    None
+                }
+            }
+        },
+        _ => None,
+    }
+}
+
+fn unit_closure<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'a hir::Expr) -> Option<(&'tcx hir::Arg, &'a hir::Expr)> {
+    if let hir::ExprClosure(_, ref decl, inner_expr_id, _, _) = expr.node {
+        let body = cx.tcx.hir.body(inner_expr_id);
+        let body_expr = &body.value;
+
+        if_chain! {
+            if decl.inputs.len() == 1;
+            if is_unit_expression(cx, body_expr);
+            if let Some(binding) = iter_input_pats(&decl, body).next();
+            then {
+                return Some((binding, body_expr));
+            }
+        }
+    }
+    None
+}
+
+/// Builds a name for the let binding variable (var_arg)
+///
+/// `x.field` => `x_field`
+/// `y` => `_y`
+///
+/// Anything else will return `_`.
+fn let_binding_name(cx: &LateContext, var_arg: &hir::Expr) -> String {
+    match &var_arg.node {
+        hir::ExprField(_, _) => snippet(cx, var_arg.span, "_").replace(".", "_"),
+        hir::ExprPath(_) => format!("_{}", snippet(cx, var_arg.span, "")),
+        _ => "_".to_string()
+    }
+}
+
+fn suggestion_msg(function_type: &str, map_type: &str) -> String {
+    format!(
+        "called `map(f)` on an {0} value where `f` is a unit {1}",
+        map_type,
+        function_type
+    )
+}
+
+fn lint_map_unit_fn(cx: &LateContext, stmt: &hir::Stmt, expr: &hir::Expr, map_args: &[hir::Expr]) {
+    let var_arg = &map_args[0];
+    let fn_arg = &map_args[1];
+
+    let (map_type, variant, lint) =
+        if match_type(cx, cx.tables.expr_ty(var_arg), &paths::OPTION) {
+            ("Option", "Some", OPTION_MAP_UNIT_FN)
+        } else if match_type(cx, cx.tables.expr_ty(var_arg), &paths::RESULT) {
+            ("Result", "Ok", RESULT_MAP_UNIT_FN)
+        } else {
+            return
+        };
+
+    if is_unit_function(cx, fn_arg) {
+        let msg = suggestion_msg("function", map_type);
+        let suggestion = format!("if let {0}({1}) = {2} {{ {3}(...) }}",
+                                 variant,
+                                 let_binding_name(cx, var_arg),
+                                 snippet(cx, var_arg.span, "_"),
+                                 snippet(cx, fn_arg.span, "_"));
+
+        span_lint_and_then(cx, lint, expr.span, &msg, |db| {
+            db.span_approximate_suggestion(stmt.span, "try this", suggestion);
+        });
+    } else if let Some((binding, closure_expr)) = unit_closure(cx, fn_arg) {
+        let msg = suggestion_msg("closure", map_type);
+
+        enum Suggestion {
+            Full(String),
+            Approx(String)
+        }
+
+        let suggestion = if let Some(expr_span) = reduce_unit_expression(cx, closure_expr) {
```

Comment:
just move this `if let` into the lint closure. That way you don't need the intermediate enum.